compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraph.kt

```diff
@@ -86,6 +86,49 @@ class ControlFlowGraph(val declaration: FirDeclaration?, val name: String, val k
     }
 }
 
+data class Edge(
+    val label: EdgeLabel,
+    val kind: EdgeKind,
+) {
+    companion object {
+        private val Normal_Forward = Edge(EdgeLabel.Normal, EdgeKind.Forward)
+        private val Normal_DeadForward = Edge(EdgeLabel.Normal, EdgeKind.DeadForward)
+        private val Normal_DfgForward = Edge(EdgeLabel.Normal, EdgeKind.DfgForward)
+        private val Normal_CfgForward = Edge(EdgeLabel.Normal, EdgeKind.CfgForward)
+        private val Normal_CfgBackward = Edge(EdgeLabel.Normal, EdgeKind.CfgBackward)
+        private val Normal_DeadBackward = Edge(EdgeLabel.Normal, EdgeKind.DeadBackward)
+
+        fun create(label: EdgeLabel, kind: EdgeKind): Edge =
+            when (label) {
+                EdgeLabel.Normal -> {
+                    when (kind) {
+                        EdgeKind.Forward -> Normal_Forward
+                        EdgeKind.DeadForward -> Normal_DeadForward
+                        EdgeKind.DfgForward -> Normal_DfgForward
+                        EdgeKind.CfgForward -> Normal_CfgForward
+                        EdgeKind.CfgBackward -> Normal_CfgBackward
+                        EdgeKind.DeadBackward -> Normal_DeadBackward
+                    }
+                }
+                else -> {
+                    Edge(label, kind)
+                }
+            }
+    }
+}
+
+enum class EdgeLabel(val label: String?) {
```

Comment:
I just realize that `EdgeLabel` can not be enum, because of labeled returns:
```kotlin
fun test(b: Boolean) {
    run {
        try {
            if (b) return
            if (b) return@run 
        } finally {
            println("Hello")
        }
    }
}
```
Here we have three kinds of edges:
- normal exit
- return from containing function
- return from lambda `run`

And there may be multiple returns from different lambdas. What do you think about making `EdgeLabel` a sealed class with two `object` implementations (for normal exit and simple return) and one class for labeled returns, which would hold symbol of targeted lambda as key?

Reply:
I agree that `enum` is quite limited, and a sealed class sounds much better!

I found that even simple return---the return from containing function---has a symbol of targeted function. So, I think `ReturnLabel` in general can be declared as a holder of that target symbol.

Instead of simple return as `object` implementation, we need uncaught exception flow as an object, and that's the whole point of this chain of commits. Thus, I added `NormalPath` and `UncaughtExceptionPath` as `object`s. `ReturnLabel` is currently not used; I prefer a manageable size of cluster that has related items. Left TODOs, and will do so afterward.

Reply:
Note that return without label path is equal to `UncaughtExceptionPath`, so there is no need create special labels for such return, only for labeled ones